<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Gloria BENOIT" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Input files - Basic tutorial for performing a GWAS on DNAnexus</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Input files";
        var mkdocs_page_input_path = "input.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Basic tutorial for performing a GWAS on DNAnexus
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../setup/">First setup</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../jobs/">About jobs</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Input files</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#genetic-data">Genetic data</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#plink2">PLINK2</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#additional-data">Additional data</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#phenotype">Phenotype</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#individual-ids">Individual ids</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#covariates">Covariates</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../plink/">Using PLINK2</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../regenie/">Using regenie</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../results/">Visualizing results</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Basic tutorial for performing a GWAS on DNAnexus</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Input files</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/gloriabenoit/DNAnexus-GWAS/edit/master/docs/input.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="input-files">Input files</h1>
<h2 id="genetic-data">Genetic data</h2>
<p>On DNAnexus, whole genome sequences data are saved in multiple formats: BGEN format (<code>.bgen</code> and <code>.sample</code>), PLINK format (<code>.bed</code>, <code>.bim</code> and <code>.fam</code>) or pVCF format (<code>.vcf.gz.tbi</code>), among others. For more information on the data available, you can run the following command:</p>
<pre><code class="language-bash">dx ls &quot;/Bulk/Whole genome sequences&quot;
</code></pre>
<h3 id="plink2">PLINK2</h3>
<p>PLINK2 can read both BGEN and PLINK format. Therefore, we have 4 ways of inputing genetic data:</p>
<p><center></p>
<table>
<thead>
<tr>
<th>File format</th>
<th>Mounted</th>
</tr>
</thead>
<tbody>
<tr>
<td>BGEN</td>
<td>No</td>
</tr>
<tr>
<td>BGEN</td>
<td>Yes</td>
</tr>
<tr>
<td>PLINK</td>
<td>No</td>
</tr>
<tr>
<td>PLINK</td>
<td>Yes</td>
</tr>
</tbody>
</table>
<p></center></p>
<p>After some trial and error, we have found that using the BGEN format without mounting it was the quickest way to perform a GWAS using PLINK2. Downloading and translating BGEN files locally was quicker than downloading or mounting the PLINK files.</p>
<p>Therefore, in this tutorial, we will input BGEN files directly. The path to the genetic data is the following: <code>"/Bulk/Whole genome sequences/Population level genome variants, BGEN format - interim 200k release/"</code>.</p>
<h2 id="additional-data">Additional data</h2>
<p>In order to run our GWAS, apart from the genetic and sample data, we need 3 additionnal files:</p>
<ul>
<li>The phenotype (for instance, BMI)</li>
<li>The ids of individuals we wish to keep (for instance, white british)</li>
<li>The covariates to use (for instance, 18 genetic principal components, the sex and the age)</li>
</ul>
<p>Theses files can either be uploaded directly to your projects using <code>dx upload</code>, or created using DNAnexus data. This tutorial will guide you through the second option.</p>
<blockquote>
<p>All three files will be uploaded to your current <code>dx</code> repertory (<code>gwas_tutorial</code> if you are closely following this tutorial).
If you want to upload them somewhere else, you can either move directory with <code>dx cd</code> before uploading, or use the <code>--path</code> or <code>--destination</code> option to specify the DNAnexus path to upload the files.</p>
</blockquote>
<h3 id="phenotype">Phenotype</h3>
<p>To extract the phenotype that we want, we first need to download all available data-fields in our dataset using Python.</p>
<blockquote>
<p>Please note, when running the <code>extract_dataset</code> command you might encounter a <code>('Invalid JSON received from server', 200)</code> error. If this happens, you simply need to rerun the code.</p>
</blockquote>
<pre><code class="language-python">&quot;&quot;&quot; Extract dataset. &quot;&quot;&quot;
import dxpy
import subprocess

dispensed_dataset_id = dxpy.find_one_data_object(typename='Dataset', name='app*.dataset', folder='/', name_mode='glob')['id']

# Get project ID
project_id = dxpy.find_one_project()[&quot;id&quot;]
dataset = (':').join([project_id, dispensed_dataset_id])

cmd = [&quot;dx&quot;, &quot;extract_dataset&quot;, dataset, &quot;-ddd&quot;, &quot;--delimiter&quot;, &quot;,&quot;]
subprocess.check_call(cmd)
</code></pre>
<p>Once completed, we want to rename the output file to a more manageable name like <code>ukbb</code>.</p>
<pre><code class="language-bash">dataset_name=$(basename *.data_dictionary.csv | sed -e &quot;s/.data_dictionary.csv//&quot;)
rename $dataset_name ukbb $dataset_name*
</code></pre>
<p>Combining these two scripts outputs 3 files:</p>
<ul>
<li><code>ukbb.data_dictionary.csv</code> contains a table with participants as the rows and metadata along the columns, including field names (see table below for naming convention)</li>
<li><code>ukbb.codings.csv</code> contains a lookup table for the different medical codes, including ICD-10 codes that will be displayed in the diagnosis field column (p41270)</li>
<li><code>ukbb.entity_dictionary.csv</code> contains the different entities that can be queried, where participant is the main entity that corresponds to most phenotype fields</li>
</ul>
<p>We first need to get the field name of the phenotype(s) we want to extract. As an example, we will extract the BMI index (<a href="https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=21001">21001</a>), but you can extract any number of phenotypes.</p>
<p>For the main participant phenotype entity, the Research Analysis Platform (UKB-RAP) uses field names with the following convention:</p>
<p><center></p>
<table>
<thead>
<tr>
<th>Type of field</th>
<th>Syntax for field name</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr>
<td>Neither instanced nor arrayed</td>
<td><code>p&lt;FIELD&gt;</code></td>
<td><code>p31</code></td>
</tr>
<tr>
<td>Instanced but not arrayed</td>
<td><code>p&lt;FIELD&gt;_i&lt;INSTANCE&gt;</code></td>
<td><code>p40005_i0</code></td>
</tr>
<tr>
<td>Arrayed but not instanced</td>
<td><code>p&lt;FIELD&gt;_a&lt;ARRAY&gt;</code></td>
<td><code>p41262_a0</code></td>
</tr>
<tr>
<td>Instanced and arrayed</td>
<td><code>p&lt;FIELD&gt;_i&lt;INSTANCE&gt;_a&lt;ARRAY&gt;</code></td>
<td><code>p93_i0_a0</code></td>
</tr>
</tbody>
</table>
<p></center></p>
<p>This means one phenotype ID can actually have multiple data field. For example, BMI has four instances.
The following python script will extract the phenotype that you wish, and every array or instance associated.</p>
<pre><code class="language-python">&quot;&quot;&quot; Extract phenotype(s) from UKBB based on field ID(s). &quot;&quot;&quot;

import os
import subprocess
import pandas as pd

# Input
FILENAME = &quot;ukbb.dataset.data_dictionary.csv&quot;
OUTPUT = &quot;pheno_extract.csv&quot;
DATASET = &quot;&lt;record-id&gt;&quot;
FIELD_ID = [21001] # BMI id

def field_names_for_ids(filename, field_ids):
    &quot;&quot;&quot; Converts data-field id to corresponding field name.

    Parameters
    ----------
    filename : str
        Path to the '.dataset.data_dictionary.csv' file.
    field_ids : list
        All field ids.
    data.frame (pandas)
        Projects's .data_dictionary.csv file.

    Returns
    -------
    list
        All corresponding field names.
    &quot;&quot;&quot;
    data = pd.read_csv(filename, sep=',')
    field_names = [&quot;eid&quot;]
    for _id in field_ids:
        select = list(data[data.name.str.match(r'^p{}(_i\d+)?(_a\d+)?$'.format(_id))].name.values)
        field_names += select

    field_names = [f&quot;participant.{f}&quot; for f in field_names]
    return field_names

# Convert id to names
FIELD_NAMES = field_names_for_ids(FILENAME, FIELD_ID)
FIELD_NAMES = &quot;,&quot;.join(FIELD_NAMES)

# Extract phenotype(s)
if os.path.exists(OUTPUT):
  os.remove(OUTPUT)
cmd = [&quot;dx&quot;, &quot;extract_dataset&quot;, DATASET, &quot;--fields&quot;, FIELD_NAMES,
        &quot;--delimiter&quot;, &quot;,&quot;, &quot;--output&quot;, OUTPUT]
subprocess.check_call(cmd)
</code></pre>
<p>This command outputs 1 file:</p>
<ul>
<li><code>pheno_extract.csv</code> contains the values for participant IDs and every single instance of all phenotypes extracted</li>
</ul>
<blockquote>
<p>Please be aware, since <code>extract_dataset</code> has no <em>overwrite</em> option by design, we implemented ourselves.
Running the previous code will first delete <code>pheno_extract.csv</code> if it's present, allowing for the extraction to happen.</p>
</blockquote>
<p>PLINK and regenie use the same formatting for the phenotype file, with a single exception : the code for missing values.
Apart from this, both need to duplicate the individuals ids and only keep the first instance for our phenotype.</p>
<p>More information on phenotype files formatting can be found <a href="https://www.cog-genomics.org/plink/2.0/input#pheno">here</a> for PLINK2 and <a href="https://rgcgithub.github.io/regenie/options/#phenotype-file-format">here</a> for regenie.</p>
<pre><code class="language-python">&quot;&quot;&quot; Format phenotype file. &quot;&quot;&quot;

import pandas as pd

# Input
FILENAME = &quot;pheno_extract.csv&quot;
PHENOTYPE = &quot;BMI&quot;
SOFTWARE = 'p' # for PLINK2 or 'r' for regenie

def format_phenotype(filename, phenotype, software):
    &quot;&quot;&quot; Save phenotype according to software format.

    Parameters
    ----------
    filename : str
        Phenotype file to format.
    phenotype : str
        Phenotype name.
    software : str { 'p', 'r'}
        Either 'p' for PLINK2 or 'r' for regenie,
        to correctly format the file.
    &quot;&quot;&quot;
    # Software specific format.
    if software == 'p':
        na_val = -9
        output = f&quot;plink_{phenotype}.txt&quot;
    elif software == 'r':
        na_val = 'NA'
        output = f&quot;regenie_{phenotype}.txt&quot;

    # Data
    data = pd.read_csv(filename, sep=',')

    # To be changed accordingly
    pheno = data.iloc[:,[0,1]]
    pheno = pheno.rename(columns={pheno.columns[0]: &quot;IID&quot;, pheno.columns[1]: phenotype})
    data = data.rename(columns={data.columns[0]: &quot;FID&quot;})
    merged = pd.concat([data.iloc[:, 0], pheno], axis=1)

    # Save output
    merged.to_csv(output, sep='\t', index=False, header=True, na_rep=na_val)

# Save phenotype
format_phenotype(FILENAME, PHENOTYPE, SOFTWARE)
</code></pre>
<p>This command outputs 1 file:</p>
<ul>
<li><code>&lt;software&gt;_BMI.csv</code> contains the formatted phenotype values</li>
</ul>
<blockquote>
<p>You can modify the <code>format_phenotype</code> function to your heart's desire, depending on what you wish to do with the phenotypes.
This is only a suggestion, and might not work for more specific phenotypes.</p>
</blockquote>
<p>You can now upload the formated phenotype file.</p>
<pre><code class="language-bash">dx upload plink_BMI.txt
dx upload regenie_BMI.txt
</code></pre>
<blockquote>
<p>For the sake of this tutorial, we upload two files: one for each type of formatting.</p>
</blockquote>
<h3 id="individual-ids">Individual ids</h3>
<p>Running a GWAS on a specific population helps reduce the bias caused by population stratification. It is therefore an important step.</p>
<p>To filter out individuals based on their ethnic background, we can use the <a href="#phenotype"><em>phenotype extraction script</em></a> and the data-field <a href="https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=21000">21000</a>. You simply need to replace the following line:</p>
<pre><code class="language-python">FIELD_ID = [21000] # ethnic background id
</code></pre>
<p>This field uses the <a href="https://biobank.ndph.ox.ac.uk/ukb/coding.cgi?id=1001">1001 data encoding</a>. In this code, <em>1001</em> means <em>white british</em> which is the main ethnic group, and the one we want to select.</p>
<pre><code class="language-bash">pop_code=1001
pop_name=&quot;white_british&quot;
population=&quot;pheno_extract.csv&quot;

awk -F &quot;,&quot; -v var=&quot;$pop_code&quot; '$2~var{print $1,$1}' $population &gt; $pop_name.txt
</code></pre>
<p>This command outputs 1 file:</p>
<ul>
<li><code>white_british.txt</code> contains the participants IDs of the white british ethnic background</li>
</ul>
<p>You can now upload the ids of your individuals.</p>
<pre><code class="language-bash">dx upload white_british.txt
</code></pre>
<h3 id="covariates">Covariates</h3>
<p>The genetic principal components from the UKBB individuals are stored in the <a href="https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=22009">22009</a> data field.
We could use our phenotype extraction algorithm, but for some reason, it is not working. Therefore, we will use another script.</p>
<p>We will use 20 variables as covariates: the first 18 PCA components, the sex (<a href="https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=31">31</a>) and the age (<a href="https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=21003">21003</a>) of individuals. However, you can extract whatever you want as covariates.</p>
<p>More information on covariates files formatting can be found <a href="https://www.cog-genomics.org/plink/2.0/input#covar">here</a> for PLINK2 and <a href="https://rgcgithub.github.io/regenie/options/#covariate-file-format">here</a> for regenie.</p>
<pre><code class="language-bash">record_id=&quot;&lt;record-id&gt;&quot;
field_names=&quot;participant.eid,participant.eid,&quot;

for i in {1..18}
do
  field_names+=&quot;participant.p22009_a$i,&quot;
done

field_names+=&quot;participant.p31,participant.p21003_i0&quot; # Sex and age

dx extract_dataset $record_id --fields $field_names --delimiter &quot;,&quot; --output covariates.txt

echo -e &quot;FID,IID,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10,PC11,PC12,PC13,PC14,PC15,PC16,PC17,PC18,Sex,Age&quot; &gt; file.tmp
tail -n+2 covariates.txt &gt;&gt; file.tmp
awk -F , '$3!=&quot;&quot;' file.tmp &gt; covariates.txt # Remove ind with no PC data
sed 's/,/\t/g' covariates.txt &gt; file.tmp
mv file.tmp covariates.txt
</code></pre>
<p>This command outputs 1 file:</p>
<ul>
<li><code>covariates.txt</code> contains the first 18 PCA components, the sex and the age of all participants</li>
</ul>
<p>You can now upload the covariates file.</p>
<pre><code class="language-bash">dx upload covariates.txt
</code></pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../jobs/" class="btn btn-neutral float-left" title="About jobs"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../plink/" class="btn btn-neutral float-right" title="Using PLINK2">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/gloriabenoit/DNAnexus-GWAS" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../jobs/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../plink/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
