<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="description" content="None" /><meta name="author" content="Gloria BENOIT" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>Basic tutorial for performing a GWAS on DNAnexus</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Home";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="." class="icon icon-home"> Basic tutorial for performing a GWAS on DNAnexus
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Home</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#structure">Structure</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#durations-and-costs">Durations and costs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#input-files">Input files</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#using-plink2">Using PLINK2</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#using-regenie">Using regenie</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#final-architecture">Final architecture</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="start/">Getting started</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="jobs/">About jobs</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="input/">Input files</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="plink/">Using PLINK2</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="regenie/">Using regenie</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="results/">Visualizing results</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href=".">Basic tutorial for performing a GWAS on DNAnexus</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Home</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/gloriabenoit/DNAnexus-GWAS/edit/master/docs/index.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="home">Home</h1>
<h2 id="overview">Overview</h2>
<p>This tutorial provides a step-by-step guide on the use of DNAnexus to perform a GWAS on the UK Biobank data.
We will run everything from the command line, but will monitor our jobs on the project's web page.</p>
<p>This tutorial assumes that you already have a researcher account on the UKBiobank site, and have access to a DNAnexus project.
If not, please create an account <a href="https://ams.ukbiobank.ac.uk/ams/signup">here</a>, and once it has been validated, create an account on <a href="https://ukbiobank.dnanexus.com/register">UKB-RAP</a> with your UK Biobank credentials to access DNAnexus. To setup your first project, please check the <a href="https://dnanexus.gitbook.io/uk-biobank-rap/getting-started/quickstart/creating-a-project">official documentation on the matter</a>.</p>
<p>This tutorial will guide you through every step needed in order to perform a basic GWAS using whole genome sequences for chromosome 1 to 22 for both <a href="https://www.cog-genomics.org/plink/2.0/">PLINK2</a> and <a href="https://rgcgithub.github.io/regenie/">regenie</a>. In order to parallelize the analyses, we will perform 22 different GWAS, one for each chromosome, and combine the results locally to reduce cost.</p>
<p>As an example, we will perform a linear regression on the <strong>BMI index</strong> (<a href="https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=21001">21001</a>) using <strong>whole genome sequencing data</strong>, specifically the <strong>interim 200k release</strong> (<code>"/Bulk/Whole genome sequences/Population level genome variants, BGEN format - interim 200k release/"</code>). However, you can use any data that you need, simply keep in mind that paths need to be changed in the scripts.</p>
<blockquote>
<p>This tutorial is written for Linux operating systems. Commands may vary accross operating systems.</p>
</blockquote>
<p>The <a href="https://dnanexus.gitbook.io/uk-biobank-rap/science-corner/gwas-using-alzheimers-disease">official GWAS guide using Alzheimer's disease</a> and the <a href="https://github.com/dnanexus/UKB_RAP/tree/main/GWAS">official github page for a GWAS on the Research Analysis Platform using regenie</a> were useful material when writing the <a href="regenie/">regenie section</a>.</p>
<p>If you want to try other tools on DNAnexus, we recommend the following github page: <a href="https://github.com/pjgreer/ukb-rap-tools">ukb-rap-tools</a> by <a href="https://github.com/pjgreer">Phil Greer</a>.</p>
<h2 id="structure">Structure</h2>
<p>The tutorial is separated into four main sections:</p>
<ol>
<li>Fondamentals for first-time use of DNAnexus (<a href="start/">Getting started</a>, <a href="jobs/">About jobs</a>)</li>
<li>Necessary files for GWAS input (<a href="input/">Input files</a>)</li>
<li>Running a GWAS (<a href="plink/">Using PLINK2</a>, <a href="regenie/">Using regenie</a>)</li>
<li>Generating plots from results (<a href="results/">Visualizing results</a>)</li>
</ol>
<blockquote>
<p>Please note, the <a href="plink/">PLINK2</a> and <a href="regenie/">regenie</a> sections are independant of each other and can be done separately. However, you will need to follow first the <a href="start/">Getting started</a> and <a href="input/">Input files</a> pages to make sure you have everything necessary to their completion.</p>
</blockquote>
<h2 id="requirements">Requirements</h2>
<p>To follow this tutorial, you will only need <a href="https://www.python.org/downloads/">Python 3</a>.</p>
<h2 id="durations-and-costs">Durations and costs</h2>
<p>Please note, when first using your account you have an initial credit of £40. Running all of this tutorial with the same instance and priority as us might go over this budget. If it is not done already, your project should be billed to a wallet which is different from your initial credit. Please check the <a href="https://documentation.dnanexus.com/admin/billing-and-account-management">official documentation</a> if you are unsure on how to proceed.</p>
<p>Jobs cost will be computed based on the official <a href="https://20779781.fs1.hubspotusercontent-na1.net/hubfs/20779781/Product%20Team%20Folder/Rate%20Cards/BiobankResearchAnalysisPlatform_Rate%20Card_Current.pdf">UKB RAP Rate Card</a> (v3.0).</p>
<blockquote>
<p>Please remember the cost and duration of a job depends on the instance and priority used.</p>
</blockquote>
<h3 id="input-files">Input files</h3>
<p>The <a href="input/">input files extraction/upload</a> is done locally and does not cost anything.</p>
<blockquote>
<p>Please be aware that storing files onto DNAnexus will result in a monthly cost. You may check its current value in the <em>SETTINGS</em> tab on your project's web page, or compute it using the <a href="https://20779781.fs1.hubspotusercontent-na1.net/hubfs/20779781/Product%20Team%20Folder/Rate%20Cards/BiobankResearchAnalysisPlatform_Rate%20Card_Current.pdf">UKB RAP Rate Card</a>.</p>
</blockquote>
<h3 id="using-plink2">Using PLINK2</h3>
<p>The <a href="plink/">PLINK2 GWAS</a> will run 22 different jobs (one per chromosome).</p>
<p>With our chosen instance (mem1_ssd1_v2_x16) using a <code>high</code> priority, the whole GWAS will take about <strong>200 minutes</strong> (3h20) and cost around <strong>£17.04</strong> (for a total execution time of 2341 minutes).</p>
<p>With the same instance (mem1_ssd1_v2_x16) using a <code>low</code> priority, if no jobs are interrupted, it will cost only <strong>£4.56</strong> for the same time.</p>
<blockquote>
<p>Please note that it is most unlikely for jobs to be uninterrupted. In our experience, the whole GWAS using this instance (mem1_ssd1_v2_x16) with a <code>low</code> priority and some interruptions has cost around <strong>£9.30</strong> altough it took almost <strong>7h30</strong> to complete (with no failed jobs).</p>
</blockquote>
<h3 id="using-regenie">Using regenie</h3>
<p>The <a href="regenie/">regenie GWAS</a> runs a total of 47 different jobs:</p>
<ul>
<li>22 jobs for the quality control (one per chromosome)</li>
<li>2 jobs for the merging and QC of merged data</li>
<li>1 job for regenie's step 1 (SNPs contribution estimation)</li>
<li>22 job for regenie's step 2 (regression, one per chromosome)</li>
</ul>
<blockquote>
<p>To optimize your GWAS's overall time, you can perform the QC in parallel to every other step beside step 2 (for which it is needed).
The rest needs to be done in order.</p>
</blockquote>
<p>With our chosen instance (mem1_ssd1_v2_x16) using a <code>high</code> priority, the whole GWAS will take about <strong>X minutes</strong> (XhY) and cost around <strong>£Y</strong> (for a total execution time of Z minutes).</p>
<p>With the same instance (mem1_ssd1_v2_x16) using a <code>low</code> priority, if no jobs are interrupted, it will cost only <strong>£Y</strong> for the same time.</p>
<p>You can find the details about the jobs run in the following table:</p>
<p><center></p>
<table>
<thead>
<tr>
<th>Action</th>
<th style="text-align: center;">Time</th>
<th style="text-align: center;">Execution time</th>
<th style="text-align: center;">Cost (high)</th>
<th style="text-align: center;">Cost (low, no interruption)</th>
</tr>
</thead>
<tbody>
<tr>
<td>QC</td>
<td style="text-align: center;">2h39</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr>
<td>Merge</td>
<td style="text-align: center;">51min</td>
<td style="text-align: center;">51min</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
<tr>
<td>Merge QC</td>
<td style="text-align: center;">10min</td>
<td style="text-align: center;">10min</td>
<td style="text-align: center;">0.07</td>
<td style="text-align: center;">0.02</td>
</tr>
<tr>
<td>Step 1</td>
<td style="text-align: center;">7h10</td>
<td style="text-align: center;">7h10</td>
<td style="text-align: center;">3.13</td>
<td style="text-align: center;">0.83</td>
</tr>
<tr>
<td>Step 2</td>
<td style="text-align: center;">48min</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
</tr>
</tbody>
</table>
<p></center></p>
<blockquote>
<p>Please note that it is most unlikely for jobs to be uninterrupted. In our experience, the whole GWAS using this instance (mem1_ssd1_v2_x16) with a <code>low</code> priority and some interruptions has cost around <strong>£Y</strong> (without accounting for failed jobs, of which there were a few). Since most steps need to be done in order, it also took a total of <strong>X</strong> to complete the GWAS (over multiple sessions since some jobs failed).</p>
</blockquote>
<h2 id="final-architecture">Final architecture</h2>
<p>At the end of this tutorial, your DNAnexus project's architecture should look like this:</p>
<pre><code class="language-text">├── Bulk
├── Showcase metadata
├── gwas_tutorial
│   ├── plink_gwas_BMI
│   │   ├── sumstat_c1.BMI.glm.linear
│   │   ├── ...
│   │   └── sumstat_c22.BMI.glm.linear
│   ├── reg_gwas_BMI
│   │   ├── merge
│   │   ├── QC_lists
│   │   ├── sumstat_c1_BMI.regenie
│   │   ├── ...
│   │   └── sumstat_c22_BMI.regenie
│   ├── covariates.txt
│   ├── plink_BMI.txt
│   ├── reg_BMI.txt
│   └── white_british.txt
├── app-id
└── app-id.dataset
</code></pre>
<blockquote>
<p>This will vary based on whether you use both PLINK2 and regenie, or only one, and if you have changed files/repertory names. Please note, if modified, files and repertory names have to be the same across all commands.</p>
</blockquote>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="start/" class="btn btn-neutral float-right" title="Getting started">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/gloriabenoit/DNAnexus-GWAS" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
    
      <span><a href="start/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="js/jquery-3.6.0.min.js"></script>
    <script>var base_url = ".";</script>
    <script src="js/theme_extra.js"></script>
    <script src="js/theme.js"></script>
      <script src="search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>

<!--
MkDocs version : 1.6.1
Build Date UTC : 2025-03-21 09:56:11.155955+00:00
-->
