<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Gloria BENOIT" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Using PLINK2 - Basic tutorial for performing a GWAS on DNAnexus</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Using PLINK2";
        var mkdocs_page_input_path = "plink.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Basic tutorial for performing a GWAS on DNAnexus
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../start/">Getting started</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../jobs/">About jobs</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../input/">Input files</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Using PLINK2</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#input-files">Input files</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#running-a-gwas">Running a GWAS</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#quality-control">Quality control</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#linear-regression">Linear regression</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#computing-the-results">Computing the results</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../regenie/">Using regenie</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../results/">Visualizing results</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Basic tutorial for performing a GWAS on DNAnexus</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Using PLINK2</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/gloriabenoit/DNAnexus-GWAS/edit/master/docs/plink.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="using-plink2">Using PLINK2</h1>
<p>This section is done from <strong>your bash terminal</strong>.
Every file created during this analysis will be stored in a main directory called <code>plink_gwas_BMI</code>.
You will run 22 different jobs (one per chromosome).</p>
<p>With our chosen instance (mem1_ssd1_v2_x16) using a <code>high</code> priority, the whole GWAS will take about <strong>40 minutes</strong> and cost around <strong>£3.84</strong> (for a total execution time of 528 minutes).</p>
<p>With the same instance (mem1_ssd1_v2_x16) using a <code>low</code> priority, if no jobs are interrupted, it will cost only <strong>£1.03</strong> for the same time.</p>
<blockquote>
<p>Please note that it is most unlikely for jobs to be uninterrupted.
In our experience, when accounting for interrupted/failed jobs, the total cost is around one to two times higher than the <em>low</em> cost, but still lower than the <em>high</em> cost. Although please keep in mind that the total time needed can be two to five times higher, depending on the interrupted/failed jobs.</p>
</blockquote>
<h2 id="input-files">Input files</h2>
<p>The path to the genetic data chosen is the following: <code>/Bulk/DRAGEN WGS/DRAGEN population level WGS variants, PLINK format [500k release]/</code>.</p>
<p>Before running a GWAS on DNAnexus using PLINK2, you need to make sure you have these 3 files uploaded to DNAnexus:</p>
<ul>
<li>The phenotype: <code>BMI.txt</code></li>
<li>The ids of individuals we wish to keep: <code>white_british.txt</code></li>
<li>The covariates to use: <code>covariates.txt</code></li>
</ul>
<p>You can check their presence with the following command:</p>
<pre><code class="language-bash">dx ls
</code></pre>
<p>Please refer to the <a href="../input/">Input files section</a> if you don't have these files.</p>
<h2 id="running-a-gwas">Running a GWAS</h2>
<p>On DNAnexus, PLINK2 is available either as part of the <a href="https://ukbiobank.dnanexus.com/app/swiss-army-knife">Swiss Army Knife app</a> (<code>swiss-army-knife</code>) or as its <a href="https://ukbiobank.dnanexus.com/app/plink_gwas">own app</a> called <code>plink_gwas</code>. We will use the former.</p>
<p>We choose to use the same instance for all GWASs, to simplify the code, but this can be changed to your liking. Same for the priority and the cost limit.</p>
<h3 id="quality-control">Quality control</h3>
<p>We perform the QC at the same time as our GWAS. The variants are filtered using the following options:</p>
<pre><code class="language-text">--maf 0.01 --hwe 1e-50 --geno 0.1
</code></pre>
<blockquote>
<p>Please change the thresholds according to your preferences, but be aware any change will modify jobs execution time, cost and size.</p>
</blockquote>
<p>Please note that we choose not to use the <code>--mind</code> filter, although it is a conventional, as it removes 88% of the samples for chromosome 18.
However, we have not checked for this behavior in other data format so it might be specific to the PLINK2 format.</p>
<h3 id="linear-regression">Linear regression</h3>
<p>Following a <a href="https://academic.oup.com/bioinformatics/article/41/3/btaf067/8008994">study by Rivas and Chang in 2025</a>, we will add the <code>qt-residualize</code> modifier to the <code>--glm</code> option, which performs a single regression on the covariates upfront, and then passes the phenotype residuals.
In our experience, this divides the execution time by 3 while yielding the same results as <code>--glm</code> alone, which is why we choose to use it.
It also has the added bonus that it scales greatly when analyzing multiple phenotypes (keeping the same execution time for more than a hundred phenotypes).</p>
<blockquote>
<p>Please be aware that this option may have limits. We recommend making sure that the results are expected.</p>
</blockquote>
<pre><code class="language-bash">pheno=&quot;BMI&quot;
pheno_path=&quot;/gwas_tutorial/$pheno.txt&quot;
ind_path=&quot;/gwas_tutorial/white_british.txt&quot;
ind=$(basename &quot;$ind_path&quot;)
cov_path=&quot;/gwas_tutorial/covariates.txt&quot;
cov=$(basename &quot;$cov_path&quot;)

instance=&quot;mem1_ssd1_v2_x16&quot;
threads=16
priority=&quot;low&quot;
cost_limit=3

dx mkdir -p plink_gwas_$pheno
dx cd plink_gwas_$pheno

for chr_num in $(seq 1 22); do
    prefix=&quot;/Bulk/DRAGEN\ WGS/DRAGEN\ population\ level\ WGS\ variants,\ PLINK\ format\ [500k\ release]//ukb24308_c${chr_num}_b0_v1&quot;
    pfile=$(basename &quot;$prefix&quot;)

    plink_command=&quot;plink2 \
                --threads $threads \
                --maf 0.01 \
                --hwe 1e-50 \
                --geno 0.1 \
                --glm 'qt-residualize' 'hide-covar' \
                --keep $ind \
                --covar $cov \
                --covar-name PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10,PC11,PC12,PC13,PC14,PC15,PC16,Age,Sex \
                --pheno $pheno.txt \
                --pfile $pfile \
                --no-psam-pheno \
                --no-input-missing-phenotype \
                --out sumstat_c${chr_num}&quot;

    dx run swiss-army-knife \
        --priority &quot;$priority&quot; --cost-limit &quot;$cost_limit&quot; \
        -icmd=&quot;$plink_command&quot; \
        --instance-type &quot;$instance&quot; \
        --name=&quot;plink_gwas_${pheno}_c${chr_num}&quot; \
        --tag=&quot;plink&quot; \
        --tag=&quot;GWAS&quot; \
        --tag=&quot;$pheno&quot; \
        --tag=&quot;c${chr_num}&quot; \
        -iin=&quot;$ind_path&quot; \
        -iin=&quot;$cov_path&quot; \
        -iin=&quot;$pheno_path&quot; \
        -iin=&quot;$prefix.pgen&quot; \
        -iin=&quot;$prefix.psam&quot; \
        -iin=&quot;$prefix.pvar&quot; \
        -y
done

dx cd ../
</code></pre>
<p>This command outputs 22 files:</p>
<ul>
<li><code>sumstat_c&lt;chrom-number&gt;.BMI.glm.linear</code> (1.26 GiB total) contains the values for the regression <em>per chromosome</em></li>
</ul>
<p>The files will be stored in the main directory, <code>plink_gwas_BMI</code>.</p>
<blockquote>
<p>Please note, the commands are the same whether your phenotype is quantitative or binary. Only the name of the output file will change (<code>.glm.linear</code> or <code>.glm.logistic.hybrid</code> for linear and logistic regression respectively).</p>
</blockquote>
<h2 id="computing-the-results">Computing the results</h2>
<p>Now that all of the summary statistics are computed, we can download them and combine them into one single file.</p>
<p>By default, PLINK2 outputs a value for each of the covariates, in addition to the global p-value. However, the <code>hide-covar</code>option we have used removes covariate-specific lines from the main report. Therefore, we only need to concatenate the results into a single file.</p>
<blockquote>
<p>Please change the value of <code>type</code> based on the regression performed: "<em>linear</em>" or "<em>logistic.hybrid</em>".</p>
</blockquote>
<pre><code class="language-bash">pheno=&quot;BMI&quot;
type=&quot;linear&quot; # to change accordingly
results_path=&quot;plink_gwas_$pheno&quot;
stat_path=&quot;plink_statistics_$pheno&quot;

mkdir -p $stat_path

for chr_num in $(seq 1 22); do
    result=&quot;sumstat_c${chr_num}.$pheno.glm.$type&quot;
    dx download &quot;$results_path/$result&quot; -o $stat_path
    if [ $chr_num -eq 1 ]; then
        head -n1 &quot;$stat_path/$result&quot; &gt; &quot;plink_sumstat_${pheno}.tsv&quot;
    fi
    tail -n +2 &quot;$stat_path/$result&quot; &gt;&gt; &quot;plink_sumstat_${pheno}.tsv&quot;
done
</code></pre>
<p>This command outputs 23 files <strong>locally</strong>:</p>
<ul>
<li><code>sumstat_c&lt;chrom-number&gt;.glm.linear</code> (1.23 Go total) contains the values for the regression <em>per chromosome</em></li>
<li><code>plink_sumstat_BMI.tsv</code> (1.2 Go) contains the concatenated values for the regression</li>
</ul>
<p>The files will be stored in a new directory named <code>plink_statistics_BMI</code>, locally this time, containing all of the summary statistics per chromosome. The combination of all of them will be located outside <code>plink_statistics_BMI</code>, at the same level, making it easier to find.</p>
<blockquote>
<p>Although the result files are big, their download should only take about a minute.</p>
</blockquote>
<p>Congratulations, you have successfully completed a GWAS using PLINK2 on DNAnexus!</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../input/" class="btn btn-neutral float-left" title="Input files"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../regenie/" class="btn btn-neutral float-right" title="Using regenie">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/gloriabenoit/DNAnexus-GWAS" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../input/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../regenie/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
