<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Gloria BENOIT" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Using PLINK2 - Basic tutorial for performing GWASs on DNAnexus</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Using PLINK2";
        var mkdocs_page_input_path = "plink.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Basic tutorial for performing GWASs on DNAnexus
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../setup/">Initial setup</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../input/">Input files</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Using PLINK2</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#input-files">Input files</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#running-a-gwas">Running a GWAS</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../errors/">Errors when running jobs</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Basic tutorial for performing GWASs on DNAnexus</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Using PLINK2</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/gloriabenoit/DNAnexus-GWAS/edit/master/docs/plink.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="using-plink2">Using PLINK2</h1>
<h2 id="input-files">Input files</h2>
<p>Before running a GWAS on DNAnexus using PLINK2, you need to make sure you have these 3 files uploaded to DNAnexus:</p>
<ul>
<li>The phenotype (BMI.txt)</li>
<li>The ids of individuals we wish to keep (white british)</li>
<li>The covariates to use (covariates.txt)</li>
</ul>
<p>You can check their presence with the following command:</p>
<pre><code class="language-bash">dx ls
</code></pre>
<p>Please refer to the <a href="../input/">Input files section</a> if you don't have these files.</p>
<h2 id="running-a-gwas">Running a GWAS</h2>
<p>On DNAnexus, PLINK2 is available as part of the Swiss Army Knife app.
We choose to use the same instance for all GWASs, to simplify the code, but this can be changed to your liking. Same for the priority and the cost limit.</p>
<pre><code class="language-bash">pheno_path=&quot;/WKD_&lt;your-name&gt;/BMI.txt&quot;
pheno=$(basename &quot;$pheno_path&quot; .txt)
ind_path=&quot;/WKD_&lt;your-name&gt;/white_british&quot;
ind=$(basename &quot;$ind_path&quot;)
cov_path=&quot;/WKD_&lt;your-name&gt;/covariates.txt&quot;
cov=$(basename &quot;$cov_path&quot;)

instance=&quot;mem2_ssd1_v2_x16&quot;
threads=16
priority=&quot;low&quot;
cost_limit=3

dx mkdir -p gwas_$pheno

for chr_num in $(seq 1 22); do
    prefix=&quot;/Bulk/Whole genome sequences/Population level genome variants, BGEN format - interim 200k release//ukb24306_c${chr_num}_b0_v1&quot;
    bgen=$(basename &quot;$prefix&quot;)
    file_ind=$(dx ls &quot;$ind_path&quot; --brief)
    file_covar=$(dx ls &quot;$cov_path&quot; --brief)
    file_pheno=$(dx ls &quot;$pheno_path&quot; --brief)
    file_sample=$(dx ls &quot;$prefix.sample&quot; --brief)
    file_bgen=$(dx ls &quot;$prefix.bgen&quot; --brief)

    plink_command=&quot;plink2 \
                --threads $threads \
                --maf 0.0001 \
                --hwe 1e-50 \
                --geno 0.1 \
                --mind 0.1 \
                --glm \
                --keep $ind \
                --covar $cov \
                --covar-name PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10,PC11,PC12,PC13,PC14,PC15,PC16,PC17,PC18,Age,Sex \
                --pheno $pheno.txt \
                --bgen $bgen.bgen ref-first \
                --sample $bgen.sample \
                --no-psam-pheno \
                --out gwas_$pheno/sumstat_c${chr_num}&quot;

    dx run swiss-army-knife \
        --priority &quot;$priority&quot; --cost-limit &quot;$cost_limit&quot; \
        -icmd=&quot;$plink_command&quot; \
        --instance-type &quot;$instance&quot; \
        --name=&quot;gwas_${pheno}_c${chr_num}&quot; \
        -iin=&quot;$ind_path&quot; \
        -iin=&quot;$cov_path&quot; \
        -iin=&quot;$pheno_path&quot; \
        -iin=&quot;$prefix.sample&quot; \
        -iin=&quot;$prefix.bgen&quot; \
        -y
done
</code></pre>
<p>This script will create a new directory named <code>gwas_&lt;phenotype&gt;</code> in the project on DNAnexus. In this folder, you will find 44 files named <code>sumstat_c&lt;num&gt;</code> (two per chromosome, the summary statistics and the log).</p>
<p>Now that all of the summary statistics are computed, we can clean them up and combine them into one clean file. We will create a new directory named <code>gwas_&lt;phenotype&gt;</code>, locally this time, with inside another directory named <code>statistics</code> containing all of the summary statistics per chromosome. The combination of all of them will be located at the same level than <code>statistics</code>, making it easier to find.</p>
<pre><code class="language-bash">pheno=&quot;BMI&quot;
type=&quot;linear&quot; # to change accordingly
output_path=&quot;gwas_$pheno&quot;
stat_path=&quot;$output_path/statistics&quot;

mkdir -p $output_path
mkdir -p $stat_path

for chr_num in $(seq 1 22); do
    result=&quot;sumstat_c${chr_num}.PHENO1.glm.$type&quot;
    dx download &quot;gwas_$pheno/$result&quot; -o $stat_path
    if [ $chr_num -eq 1 ]; then
        head -n1 &quot;$stat_path/$result&quot; &gt; &quot;$output_path/sumstat_${pheno}.ADD&quot;
    fi
    head -n1 &quot;$stat_path/$result&quot; &gt; &quot;$stat_path/sumstat_c${chr_num}.ADD&quot;
    grep &quot;ADD&quot; &quot;$stat_path/$result&quot; &gt;&gt; &quot;$stat_path/sumstat_c${chr_num}.ADD&quot;
    grep &quot;ADD&quot; &quot;$stat_path/$result&quot; &gt;&gt; &quot;$output_path/sumstat_${pheno}.ADD&quot;
done
</code></pre>
<p>Congratulations, you have successfully completed a GWAS using PLINK2 on DNAnexus!</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../input/" class="btn btn-neutral float-left" title="Input files"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../errors/" class="btn btn-neutral float-right" title="Errors when running jobs">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/gloriabenoit/DNAnexus-GWAS" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../input/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../errors/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
